core/analysis/collation_token_stream.cpp

```diff
@@ -0,0 +1,354 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2021 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrey Abramov
+////////////////////////////////////////////////////////////////////////////////
+
+#include "collation_token_stream.hpp"
+
+#include <unicode/locid.h>
+#include <unicode/coll.h>
+
+#include "velocypack/Slice.h"
+#include "velocypack/Builder.h"
+#include "velocypack/Parser.h"
+#include "velocypack/velocypack-aliases.h"
+
+#include "utils/locale_utils.hpp"
+#include "utils/vpack_utils.hpp"
+
+namespace iresearch {
+namespace analysis {
+
+struct collation_token_stream::state_t {
+  icu::Locale icu_locale;
+  const options_t options;
+  std::unique_ptr<icu::Collator> collator;
+  std::string utf8_buf;
+  byte_type term_buf[32768];
+
+  state_t(const options_t& opts)
+    : icu_locale("C"), options(opts) {
+    // NOTE: use of the default constructor for Locale() or
+    //       use of Locale::createFromName(nullptr)
+    //       causes a memory leak with Boost 1.58, as detected by valgrind
+    icu_locale.setToBogus(); // set to uninitialized
+  }
+};
+
+} // analysis
+} // iresearch
+
+namespace {
+
+using namespace irs;
+
+bool make_locale_from_name(const string_ref& name,
```

Comment:
Maybe we should move this to some shared header? This is 4th instance of this function in our codebase 😢